protobuf/lib/src/protobuf/field_type.dart

```diff
@@ -5,66 +5,187 @@
 // ignore_for_file: constant_identifier_names,non_constant_identifier_names
 part of protobuf;
 
-bool _isRepeated(int fieldType) => (fieldType & PbFieldType._REPEATED_BIT) != 0;
+class FieldType {
+  final FieldBaseType baseType;
+  final bool isGroup;
+  final bool isPacked;
+  final bool isRepeated;
+  final bool isRequired;
 
-bool _isRequired(int fieldType) => (fieldType & PbFieldType._REQUIRED_BIT) != 0;
+  const FieldType.dummy()
+      : baseType = FieldBaseType.int32,
+        isGroup = false,
+        isPacked = false,
+        isRepeated = false,
+        isRequired = false;
 
-bool _isEnum(int fieldType) =>
-    PbFieldType._baseType(fieldType) == PbFieldType._ENUM_BIT;
+  const FieldType.optional(this.baseType)
+      : isGroup = false,
+        isPacked = false,
+        isRepeated = false,
+        isRequired = false;
 
-bool _isBytes(int fieldType) =>
-    PbFieldType._baseType(fieldType) == PbFieldType._BYTES_BIT;
+  const FieldType.required(this.baseType)
+      : isGroup = false,
+        isPacked = false,
+        isRepeated = false,
+        isRequired = true;
 
-bool _isGroupOrMessage(int fieldType) =>
-    (fieldType & (PbFieldType._GROUP_BIT | PbFieldType._MESSAGE_BIT)) != 0;
+  const FieldType.repeated(this.baseType)
+      : isGroup = false,
+        isPacked = false,
+        isRepeated = true,
+        isRequired = false;
 
-bool _isMapField(int fieldType) => (fieldType & PbFieldType._MAP_BIT) != 0;
+  const FieldType.packed(this.baseType)
+      : isGroup = false,
+        isPacked = true,
+        isRepeated = true, // packed implies repeated
+        isRequired = false;
+
+  const FieldType.fromBaseType(this.baseType)
+      : isGroup = false,
+        isPacked = false,
+        isRepeated = false,
+        isRequired = false;
+
+  const FieldType.MAP()
```

Comment:
It might be good to align names with Dart style guide, e.g. `MAP` should probably be `map` etc.

Even better pattern could be:

```dart
const FieldType._({
    required this.baseType,
    this.isGroup = false,
    this.isPacked = false,
    this.isRepeated = false,
    this.isRequired = false,
});

static const map = FieldType._(FieldBaseType.map); 
```

Reply:
> It might be good to align names with Dart style guide

We should check this automatically somehow, but I don't know if it's possible, and if it is how. I'll figure it out.

Reply:
I would have expected that this is caught by https://dart-lang.github.io/linter/lints/non_constant_identifier_names.html ?

It part of core and recommended lint lists. I am surprised that this is not detected. /cc @pq

Reply:
🤔 

```dart
 const FieldType.MAP()
```

should definitely get flagged by [`non_constant_identifier_names`](https://dart-lang.github.io/linter/lints/non_constant_identifier_names.html).

Lints are a little subtle in some IDEs so maybe it was just overlooked.

Here's how it should look in IntelliJ for example:


<img width="268" alt="image" src="https://user-images.githubusercontent.com/67586/164102374-1faab0ac-c4ac-44c6-87a5-a4af9e548d53.png">

Reply:
I just realized that we disable this lint at the beginning of this file.

Most of the identifiers in the files are used by the generated code and have cryptic names and I guess we disable the lints for those. For now I will fix naming of the identifiers for user-written code manually.